### PR TITLE
Hydjet2 bugfixes

### PR DESCRIPTION
**Pull request by wouf**
#### PR description:

This is bugfixes for Hydjet2 related to status codes and units issues. The default parameters was updated to Antares_pre tune

#### PR validation:

workflow 160

### CODE REVIEW

**Comment by SiewYan** [comment]
please test workflow 160

File: GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc [hunk 1]
```diff
@@ -490,10 +499,12 @@ HepMC::GenVertex *Hydjet2Hadronizer::build_hyjet2_vertex(int i, int id) {
   // build verteces for the hyjets stored events
   double x0 = (hj2->GetX()).at(i);
   double y0 = (hj2->GetY()).at(i);
-  double x = x0 * cosphi0_ - y0 * sinphi0_;
-  double y = y0 * cosphi0_ + x0 * sinphi0_;
-  double z = (hj2->GetZ()).at(i);
-  double t = (hj2->GetT()).at(i);
+
+  // convert to mm (as in PYTHIA6)
+  double x = 0.000000000001 * (x0 * cosphi0_ - y0 * sinphi0_);
```
Comment (perrotta) [inline GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc:504]: It is less error prone, and more intuitive, if you define the conversion in a contexpr parameter

File: GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc [hunk 2]
```diff
@@ -524,22 +582,27 @@ void Hydjet2Analyzer::analyze(const edm::Event &iEvent, const edm::EventSetup &i
...
-	  }else if(!(pdt.particle(pdg_id))){ //skip if not in PDT!!!
-	    cout<<" Out of PDT: "<< pdg_id<<endl;
-	    continue;
-          }else{
-	    part = pdt.particle(pdg_id);
+          if (mass < 0) {
+            if ((abs(pdg_id) == 130) || (abs(pdg_id) == 310)) {
+              mass = 0.497611;
+            } else {
+              //cout<<"Error! Mass for "<< pdg_id <<" not found in PDT!!!"<<endl;
```
Comment (perrotta) [inline GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc:595]: Does it makes sense to continue if there is a negative mass here?

File: GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc [hunk 3]
```diff
@@ -570,25 +633,119 @@ void Hydjet2Analyzer::analyze(const edm::Event &iEvent, const edm::EventSetup &i
...
           if (doHistos_) {
             //common histos
-            if ((*it)->status() < 10) {  //status 1
+            if ((*it)->status() == 1) {
+              dhpdg->Fill(pdg_id);
               dheta->Fill(eta);
 
+              e_sum += e;
+              et_sum += et;
+              if (std::abs(eta) < 1.5) {
```
Comment (perrotta) [inline GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc:642]: if you define (e.g.) `etaMaxBarrel` , `etaMaxAlice` , `etaMaxCms` it is easier to code, and modify afterwards would you ever need it

**Review by perrotta — COMMENTED** [review]
Besides the bug fixes, you are also modifying several generator parameters in hydjet2DefaultParameters_cff.py
Please add it to the PR description, so that it can remain track of it and its purposes

File: GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc (hunk 1 above)
Comment (wouf) [reply GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc:504]: How it is possible? We do not know the x0 and y0 at compile time...

File: GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc (hunk 2 above)
Comment (wouf) [reply GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc:595]: We using mass only for ET, so with negative mass we still can use other histograms. Anyway this analyser is used for tests and tuning only.

File: GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc (hunk 3 above)
Comment (wouf) [reply GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc:642]: Well, firstly we have to change detector then)) I thing it is useless to introduce variables like etaMaxBarrel. I prefer to leave it hardcoded in default section, to vary cuts from the configs, it is found userHistos_ section, see [this line](https://github.com/cms-sw/cmssw/blob/a3b55905708b74b0fdc0b65f6484d58d51310e99/GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc#L762)

File: GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc (hunk 1 above)
Comment (perrotta) [reply GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc:504]: e.g.
```C++
  double constexpr am_to_mm = 0.000000000001;  // or even 1e-12, which is more intuitive
  double x =  am_to_mm * (x0 * cosphi0_ - y0 * sinphi0_);
```

Comment (wouf) [reply GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc:504]: Ah, I see. But what is the reason of constexpr here? Why not just:
```
const double fm_to_mm = 1e-12; 
double x =  fm_to_mm * (x0 * cosphi0_ - y0 * sinphi0_);
```
?

**Comment by SiewYan** [comment]
please test

**Comment by SiewYan** [comment]
> Besides the bug fixes, you are also modifying several generator parameters in hydjet2DefaultParameters_cff.py Please add it to the PR description, so that it can remain track of it and its purposes

Hello @wouf , sorry for the delay... Could you provide the reason why those changes.

**Comment by wouf** [comment]
> > Besides the bug fixes, you are also modifying several generator parameters in hydjet2DefaultParameters_cff.py Please add it to the PR description, so that it can remain track of it and its purposes
> 
> Hello @wouf , sorry for the delay... Could you provide the reason why those changes, or , could you quote the Antares_pre tune? Thanks

This is a new version of Hydjet++ and for this moment we are tuning it. Antares pre (preliminary) tune is based on Hydjet 1.9 tune (Drum5F) which used in CMS HIN. Later we will update it using comparison with data.

**Comment by Saptaparna** [comment]
+1

**Comment by perrotta** [comment]
please test workflow 160
(to refresh them)

File: GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc (hunk 1 above)
Comment (perrotta) [reply GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc]: > Ah, I see. But what is the reason of constexpr here? Why not just:
> 
> ```
> const double fm_to_mm = 1e-12; 
> double x =  fm_to_mm * (x0 * cosphi0_ - y0 * sinphi0_);
> ```
> 
> ?

Well, `constexpr` is initialized at compile time, while `const` it is at runtime. Not a big difference here, let it as you updated it.

**Comment by perrotta** [comment]
@Saptaparna there are quite large changes in wf 160. Since you signed this when the test results were not available yet, could you please confirm that all these changes are as expected, due to the update of the default parameters to Antares_pre tune implemented in this PR?

**Comment by SiewYan** [comment]
Hello @perrotta , echoing from https://github.com/cms-sw/cmssw/pull/38443#issuecomment-1182174838, I think the large changes are expected as @wouf mentioned the Antares pre (preliminary) tune. I suggest to merge it for now. What do you think @Saptaparna ?

**Comment by qliphy** [comment]
@wouf @Saptaparna  Please confirm whether large changes in wf 160 are expected or not.  Thanks!

**Comment by wouf** [comment]
> @wouf @Saptaparna Please confirm whether large changes in wf 160 are expected or not. Thanks!

What You mean by large changes? No changes in wf within this PR found. Or You mean changes in  #PR36316 , in files 
```
Configuration/PyReleaseValidation/python/relval_standard.py
Configuration/PyReleaseValidation/python/relval_steps.py
```
? If so, i guess, @makortel can comment used conditions for wf 160.

**Comment by qliphy** [comment]
@wouf  You can find from above PR test outcome: "DQMHistoTests: Total failures: 28426"

For more details, please check here:

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-07-24-2300+e20d13/51778/validateJR.html

https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-07-24-2300+e20d13/51778/160.0_Hydjet2Q_MinBias_5020GeV_2018_ppReco+Hydjet2Q_MinBias_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECOMB+ALCARECOHI2018PPRECO+HARVESTHI2018PPRECO/

**Comment by wouf** [comment]
> @wouf You can find from above PR test outcome: "DQMHistoTests: Total failures: 28426"
> 
> For more details, please check here:
> 
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-07-24-2300+e20d13/51778/validateJR.html
> 
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-07-24-2300+e20d13/51778/160.0_Hydjet2Q_MinBias_5020GeV_2018_ppReco+Hydjet2Q_MinBias_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECOMB+ALCARECOHI2018PPRECO+HARVESTHI2018PPRECO/

I see, thanks!

@mandrenguyen Could You please to look throw it and to check if wf 160 is correct?

**Comment by qliphy** [comment]
hold 
to understand the difference on wf 160

**Comment by mandrenguyen** [comment]
Apologies, I missed that this was waiting for my attention.  Wf 160 is exactly the same as wf 158, expect with the new version of Hydjet instead of the old one. 
There is no issue with wf 160 per se. 

Hydjet is the baseline generator for PbPb events, it's hugely important for the heavy-ion program.
Hydjet++ is a completely rewritten version of this generator, and we're still trying to understand the performance. 
IIUC this PR has some genuine bugfixes, as well as a new tuning.  Big differences are expected, but Andrey should comment whether these changes go in the expected direction.
 
Note that even after this PR I don't believe the new generator is necessarily ready for production.  There will likely be further tuning necessary, and possibly further bugfixes.

**Comment by wouf** [comment]
Thank You @mandrenguyen . @qliphy , I think we can go ahead, thanks!

**Comment by qliphy** [comment]
unhold
thanks!

**Comment by qliphy** [comment]
please test
to refresh the results

**Comment by qliphy** [comment]
+1

**Pull request merged**